src/acl_kernel_if.cpp

```diff
@@ -874,10 +874,26 @@ int acl_kernel_if_update(const acl_device_def_autodiscovery_t &devdef,
     // Allocations for each kernel
     kern->accel_csr = (acl_kernel_if_addr_range *)acl_malloc(
         kern->num_accel * sizeof(acl_kernel_if_addr_range));
+    if (!(kern->accel_csr)) {
+      ACL_KERNEL_IF_DEBUG_MSG(kern,
+                              "Failed to allocate memory for kernel csr.\n");
+      return 1;
+    }
     kern->accel_perf_mon = (acl_kernel_if_addr_range *)acl_malloc(
         kern->num_accel * sizeof(acl_kernel_if_addr_range));
+    if (!(kern->accel_perf_mon)) {
+      ACL_KERNEL_IF_DEBUG_MSG(
+          kern, "Failed to allocate memory for kernel performance monitor.\n");
+      return 1;
```

Comment:
The `return` leaks memory allocated previously, e.g., for `kern->accel_csr`.

There are two options here:
1. You can either clean up the resources meticulously before each return. In C programming, this is usually implemented by having a resource clean-up block at the end of the function, and jumping to the appropriate line within that block using `goto`. In C++ programming, this is resolved using [RAII](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization), where the resource is managed by an object with a deconstructor freeing the resource.

2. Instead, I would suggest to simply abort with `assert()`. In an out-of-memory situation, it is usually difficult to recover gracefully anyway, so aborting does not make things any worse.

Reply:
Thanks, I have changed all if statements to asserts instead.